### PR TITLE
Scoping Event Broker and Changes from Urs

### PR DESCRIPTION
**Pull request by danielmarbach**
Here the pull request with the scoping event broker. Please start reviewing

### CODE REVIEW

**Comment by danielmarbach** [comment]
Documentation is still work in progress but I will send a pullrequest in the next few days

**Comment by ursenzler** [comment]
Pull request cannot be merged, please update your "head" to latest commit of appccelerate:master

**Comment by danielmarbach** [comment]
Hmmm... I did that... Let me check

Am 03.06.2013 um 09:19 schrieb Urs Enzler notifications@github.com:

**Comment by danielmarbach** [comment]
I merged the latest changes

**Comment by ursenzler** [comment]
- [x] Subject attributes are missing in specs.

**Comment by danielmarbach** [comment]
Done

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    using FluentAssertions;
+
+    using Xunit;
+    using Xunit.Extensions;
+
+    public class EventScopeFactoryTest
+    {
+        [Theory]
+        [ClassData(typeof(EventScopeFactoriesAndTheirScopes))]
+        public void CreateHandlerDecorator_ShouldCreateDefaultDecorator(IEventScopeFactory factory, Type scopeType)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: According to new unit test naming convention, the name of the unit test method should be: CreatesDefaultDecorator

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            IHandler decorator = factory.CreateHandlerDecorator(A.Fake<IHandler>());
+
+            decorator.Should().BeOfType<ScopingHandlerDecorator>();
+        }
+
+        [Theory]
+        [ClassData(typeof(EventScopeFactoriesAndTheirScopes))]
+        public void CreateScopeHolder_ShouldCreateScopeHolderOfType(IEventScopeFactory factory, Type scopeType)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: CreatesScopeHolder

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            IEventScopeHolder scopeHolder = factory.CreateScopeHolder();
+
+            scopeHolder.GetType().Should().Be(scopeType);
+        }
+
+        [Theory]
+        [ClassData(typeof(EventScopeFactoriesAndTheirScopes))]
+        public void CreateScopeContext_ShouldCreateScopeHolderOfType(IEventScopeFactory factory, Type scopeType)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: CreatesScopeContext

Also there was a copy/paste error

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs (hunk 1 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: We did never actually agree on that one. There are a lot of tests which are inconsistent. I certainly can change that if you want…

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            IEventScopeContext scopeContext = factory.CreateScopeContext();
+
+            scopeContext.GetType().Should().Be(scopeType);
+        }
+
+        [Theory]
+        [ClassData(typeof(EventScopeFactoriesAndTheirScopes))]
+        public void ScopeHolderAndScopeContext_ShouldBeSameInstance(IEventScopeFactory factory, Type scopeType)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: Why should it be the same?

Update name according to convention.

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            var scopeContext = factory.CreateScopeContext();
+            var scopeHolder = factory.CreateScopeHolder();
+
+            scopeContext.Should().BeSameAs(scopeHolder);
+        }
+
+        [Theory]
+        [ClassData(typeof(EventScopeFactoriesAndTheirScopes))]
+        public void CreateScope_ShoudlCreateDefaultEventScope(IEventScopeFactory factory, Type scopeType)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: name...

File: source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+namespace Appccelerate.ScopingEventBroker
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Reflection;
+    using Appccelerate.EventBroker;
+    using Appccelerate.EventBroker.Handlers;
+    using Appccelerate.EventBroker.Internals.Subscriptions;
+    using Appccelerate.ScopingEventBroker.Internals;
+    using Appccelerate.ScopingEventBroker.Internals.Context;
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs]: using is not needed

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs (hunk 3 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: No this is not a copy paste error. Look closely. One is the scope holder the other one is the scope context

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs (hunk 4 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs]: The context is also a scope holder. There should only be one created by the factory.

**Comment by danielmarbach** [comment]
Pushed review findings

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+    using FluentAssertions;
+
+    using Xunit.Extensions;
+
+    public class EventScopeFactoryTest
+    {
+        [Theory]
+        [ClassData(typeof(EventScopeFactoriesAndTheirScopes))]
+        public void CreateHandlerDecorator_CreatesDefaultDecorator(IEventScopeFactory factory, Type scopeType)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs:39]: You understood me wrong: name convention does not anymore contain the Feature part. It's BEHAVIOR[_WhenSCENARIO]

For this method the name should be "CreatesDefaultDecorator"

Applies to all other test methods, too.

Sorry :-)

File: source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    using Xunit;
+    using Xunit.Extensions;
+
+    public class EventScopingStandardFactoryTest
+    {
+        private const bool Decorated = true;
+        private const bool NotDecorated = false;
+
+        [Fact]
+        public void CreateScopeContext_ShouldDelegateToDecorated()
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs:38]: Change all unit test method names to match coding convention.

File: source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+            result.Should().BeSameAs(scopeContext);
+        }
+
+        [Theory]
+        [InlineData(typeof(OnBackground), Decorated)]
+        [InlineData(typeof(OnUserInterface), NotDecorated)]
+        [InlineData(typeof(OnUserInterfaceAsync), Decorated)]
+        [InlineData(typeof(OnPublisher), NotDecorated)]
+        public void CreateHandler_ShouldDecorateWhenHandlerAsynchronous(Type handlerType, bool decorated)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs]: Split test into set for synchronous and a set for asynchronous test data. Then, you don't have to write the ShouldBe method as complicated at it is now.

New test names would be:
1) DecoratesAsynchronousHandlerWithScopingHandler for asynchronous handlers
2) CreatesStandardSynchronousHandlers for synchronous handlers

File: source/Appccelerate.ScopingEventBroker.Test/Internals/Context/AbstractEventScopeContextTest.cs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        public AbstractEventScopeContextTest()
+        {
+            this.eventScope = A.Fake<IEventScopeInternal>();
+            this.eventScopeFactory = A.Fake<IEventScopeFactory>();
+
+            this.testee = new TestAbstractEventScopeContext(this.eventScopeFactory);
+        }
+
+        [Fact]
+        public void Current_WhenConstructed_ShouldBeNull()
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/Internals/Context/AbstractEventScopeContextTest.cs:46]: change name to match naming convention

Applies to all other test method names, too

File: source/Appccelerate.ScopingEventBroker.Test/Internals/Context/PerCallEventScopeContextTest.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        private readonly PerCallEventScopeContext testee;
+
+        public PerCallEventScopeContextTest()
+        {
+            this.testee = new PerCallEventScopeContext(A.Fake<IEventScopeFactory>());
+        }
+
+        [Fact]
+        public void Acquire_WhenAcquiredWithoutDispose_ShouldReturnSame()
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/Internals/Context/PerCallEventScopeContextTest.cs:37]: Naming

File: source/Appccelerate.ScopingEventBroker.Test/Internals/Context/PerTransactionScopeContextTest.cs [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        public PerTransactionScopeContextTest()
+        {
+            this.eventScope = A.Fake<IEventScopeInternal>();
+            this.eventScopeFactory = A.Fake<IEventScopeFactory>();
+
+            this.testee = new PerTransactionScopeContext(this.eventScopeFactory);
+        }
+
+        [Fact]
+        public void Acquire_WithoutTransactionScope_ShouldReturnNull()
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/Internals/Context/PerTransactionScopeContextTest.cs:47]: naming

File: source/Appccelerate.ScopingEventBroker.Test/Internals/EventScopeTest.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        private readonly EventScope testee;
+
+        public EventScopeTest()
+        {
+            this.testee = new EventScope();
+        }
+
+        [Fact]
+        public void Release_WhenNotCanceled_ShouldExecutedCallbacks()
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker.Test/Internals/EventScopeTest.cs:37]: naming

File: source/Appccelerate.ScopingEventBroker/Appccelerate.ScopingEventBroker.nuspec [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    <title>Appccelerate.ScopingEventBroker</title>
+    <authors>Appccelerate team</authors>
+    <owners>Appccelerate team</owners>
+    <licenseUrl>http://www.apache.org/licenses/LICENSE-2.0.html</licenseUrl>
+    <projectUrl>https://github.com/appccelerate/appccelerate</projectUrl>
+    <iconUrl>https://github.com/appccelerate/appccelerate/raw/master/docs/nuget.png</iconUrl>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+	<summary>Scoping event broker allows to cancel asynchronous events.</summary>
+    <description>The scoping event broker adds a notion of an unit of work to your event broker usage which allows to cancel asynchronous events.</description>
+    <copyright>Copyright 2012</copyright>
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Appccelerate.ScopingEventBroker.nuspec]: 2013

File: source/Appccelerate.ScopingEventBroker/Internals/Context/AbstractEventScopeFactory.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        public virtual IHandler CreateHandlerDecorator(IHandler handler)
+        {
+            return new ScopingHandlerDecorator(handler, this.CreateScopeHolder());
+        }
+
+        protected abstract IEventScopeContextInternal CreateScope(IEventScopeFactory eventScopeFactory);
+
+        private IEventScopeContextInternal GetOrCreate()
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/AbstractEventScopeFactory.cs]: Get or create what?
--> GetOrCreateScopeContext

File: source/Appccelerate.ScopingEventBroker/Internals/Context/IEventScopeContextInternal.cs [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+// </copyright>
+//-------------------------------------------------------------------------------
+
+namespace Appccelerate.ScopingEventBroker.Internals.Context
+{
+    public interface IEventScopeContextInternal : IEventScopeContext, IEventScopeHolder
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/IEventScopeContextInternal.cs:21]: This name hides the fact that it is also a holder. And I don't like Internal :-)

File: source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs (hunk 7 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopeFactoryTest.cs:39]: Yeah but the factory has different behaviors :)

Am 04.06.2013 um 10:43 schrieb Urs Enzler notifications@github.com:

File: source/Appccelerate.ScopingEventBroker/Internals/Context/PerTransactionScopeContext.cs [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+                return (IEventScopeInternal)this.Acquire();
+            }
+        }
+
+        public IEventScope Acquire()
+        {
+            SinglePhaseScopeDecorator scope = null;
+            Transaction transaction = Transaction.Current;
+
+            if (transaction != null && !this.scopesToTransactionIdentifier.TryGetValue(
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/PerTransactionScopeContext.cs:50]: Can this check be put into a well named method?

File: source/Appccelerate.ScopingEventBroker/Internals/Context/PerTransactionScopeContext.cs [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+                    transaction.TransactionInformation.LocalIdentifier, out scope))
+            {
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/PerTransactionScopeContext.cs:52]: Extract body of if into dedicated method?

File: source/Appccelerate.ScopingEventBroker/Internals/Context/AbstractEventScopeFactory.cs (hunk 15 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/AbstractEventScopeFactory.cs]: When there is one it should return else create

Am 04.06.2013 um 11:05 schrieb Urs Enzler notifications@github.com:

File: source/Appccelerate.ScopingEventBroker/Internals/Context/IEventScopeContextInternal.cs (hunk 16 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/IEventScopeContextInternal.cs:21]: Better name?

IEventScopeContextWithHolder?

Am 04.06.2013 um 11:07 schrieb Urs Enzler notifications@github.com:

File: source/Appccelerate.ScopingEventBroker/Internals/EventScope.cs [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        public void Register(Action releaseCallback)
+        {
+            this.callbacks.Add(releaseCallback);
+        }
+
+        public void Release()
+        {
+            foreach (Action callback in this.callbacks)
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/EventScope.cs]: Is it safe to say that a callback will not modify the callbacks collection? If not, copy the collection before iterating.

File: source/Appccelerate.ScopingEventBroker/Internals/IEventScopeInternal.cs [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+//   limitations under the License.
+// </copyright>
+//-------------------------------------------------------------------------------
+
+namespace Appccelerate.ScopingEventBroker.Internals
+{
+    /// <summary>
+    /// Marker interface which implements <see cref="IEventScope"/> and <see cref="IEventScopeRegistry"/>
+    /// </summary>
+    public interface IEventScopeInternal : IEventScope, IEventScopeRegistry
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/IEventScopeInternal.cs:24]: Hmm, again a difficult naming problem

File: source/Appccelerate.ScopingEventBroker/Version.txt [hunk 21]
```diff
@@ -0,0 +1 @@
+0.1
```
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Version.txt:1]: 1.0 ?

**Comment by ursenzler** [comment]
Great stuff!

File: source/Appccelerate.ScopingEventBroker/Version.txt (hunk 21 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker/Version.txt:1]: Ok for me

Am 04.06.2013 um 11:15 schrieb Urs Enzler notifications@github.com:

File: source/Appccelerate.ScopingEventBroker/Internals/Context/AbstractEventScopeFactory.cs (hunk 15 above)
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/AbstractEventScopeFactory.cs]: Yes, I just meant to add ScopeContext to the name so the name  of the method states what it gets or creates ;-)

File: source/Appccelerate.ScopingEventBroker/Internals/Context/IEventScopeContextInternal.cs (hunk 16 above)
Comment (ursenzler) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/IEventScopeContextInternal.cs:21]: IEventScopeContextAndHolder ?

I really don't know.

**Comment by ursenzler** [comment]
Dani, it's okay for me to merge it into appccelerate:master. 
Do it when you fixed the things worth fixing.

File: source/Appccelerate.ScopingEventBroker/Internals/EventScope.cs (hunk 19 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker/Internals/EventScope.cs]: It used to be safe without the TransactionScope stuff. But now I have to copy the collection

File: source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs (hunk 9 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker.Test/EventScopingStandardFactoryTest.cs]: Make sense and is much more clear

File: source/Appccelerate.ScopingEventBroker/Internals/Context/PerTransactionScopeContext.cs (hunk 17 above)
Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/PerTransactionScopeContext.cs:50]: I thought about that too when I implemented it but then I thought that it is a simple condition and the scopeToTransactionIdentifier dictionary is a central part of the behavior of that class so I don't want to hide it in a method but certainly a matter of style

Comment (danielmarbach) [inline source/Appccelerate.ScopingEventBroker/Internals/Context/PerTransactionScopeContext.cs:50]: And there is also the out parameter which would be added to that method which I dislike

**Comment by danielmarbach** [comment]
I pushed some of the review findings. Still missing the test method renaming!
- [ ] Rename Test method to follow convention

**Pull request merged**